### PR TITLE
Build documentation in Nix workflow.

### PR DESCRIPTION
**Pull request by Zimmi48**
Reopening #61.

### CODE REVIEW

**Comment by Zimmi48** [comment]
References show up as `[?]`. Any idea what that could be?

**Comment by Zimmi48** [comment]
Less than 16 minutes for the total duration of the Nix CI including the documentation job (which takes 9 minutes). That's 10 minutes less than the doc pipeline.

File: .github/workflows/nix-action-default.yml [hunk 1]
```diff
@@ -284,6 +284,73 @@ jobs:
...
+        nix-shell --run "cachedMake"
+        nix-shell --run "make -C doc/movies -j $(nproc) all"
+    - name: Build coqdoc
+      run: nix-shell --run "make -j $(nproc) html"
+    - name: Compile LaTeX document
+      run: |
+        nix-shell --run "make pdf"
+        # I need to run make pdf twice to get reference resolved :(
+        touch doc/hydras.tex
+        nix-shell --run "make pdf"
```
Comment (Zimmi48) [inline .github/workflows/nix-action-default.yml:332]: I run `make pdf` twice to work around the unresolved reference issue. There is probably a better fix (which probably involves fixing a makefile). Suggestions welcome!

**Comment by Zimmi48** [comment]
> Less than 16 minutes for the total duration of the Nix CI including the documentation job (which takes 9 minutes). That's 10 minutes less than the doc pipeline.

Actually, we are even faster at the second run because all the dependencies (in particular SerAPI and its own dependencies) are already built and in cache already. Now, the total duration of the pipeline is 14m40s.

There is still an issue though, which makes the pipeline slower than it should, which is that `make html` isn't compatible with `cachedMake` (probably because of timestamps being wrong) and therefore vo files are rebuilt in the `build-doc` job. We can fix this issue by moving the coqdoc build to the previous step (building `hydra-battles-single`).

**Comment by Casteran** [comment]
Hi Théo,

  It looks to be the place of the « bibtex  hydras » line in doc/Makefile.
  I made some tests on my machine
  a) remove doc/hydras.bbl
  b) make pdf
   It results in unresolved citations
  Then I changed the place of « bibles hydras » between the first and the second « lualatex hydras » 
  Removed again hydras.bbl, then it seems to work.

  I didn’t notice this issue, probably because of the old hydras.bbl from older compilations.

Could you check if changing the makefile works ?

Pierre

hydras.pdf: ${DEP_FILES}
	@+${MAKE} -C movies all
	lualatex hydras.tex
	bibtex hydras
	lualatex hydras.tex
	makeindex coq
	makeindex maths
	makeindex hydras
	makeindex primrec
	makeindex additions
	lualatex hydras.tex
	lualatex hydras.tex

File: .github/workflows/nix-action-default.yml (hunk 1 above)
Comment (Casteran) [reply .github/workflows/nix-action-default.yml:330]: @Zimmi48 Do you still need to run `make pdf`twice ? I commited the move of `bibtex hydras` upwards.

Comment (Zimmi48) [reply .github/workflows/nix-action-default.yml:330]: Cool! I'll have a look tonight.

Comment (Zimmi48) [reply .github/workflows/nix-action-default.yml]: Indeed, it works now!

**Comment by Zimmi48** [comment]
> There is still an issue though, which makes the pipeline slower than it should, which is that `make html` isn't compatible with `cachedMake` (probably because of timestamps being wrong) and therefore vo files are rebuilt in the `build-doc` job. We can fix this issue by moving the coqdoc build to the previous step (building `hydra-battles-single`).

I've done that and now `hydra-battles-single` takes much longer to build, to the point that the overall pipeline duration is bigger by a few minutes. I don't understand what is going on. If I don't manage to solve this issue, I'll revert to the version without this change.

**Comment by Zimmi48** [comment]
> There is still an issue though, which makes the pipeline slower than it should, which is that `make html` isn't compatible with `cachedMake` (probably because of timestamps being wrong) and therefore vo files are rebuilt in the `build-doc` job. We can fix this issue by moving the coqdoc build to the previous step (building `hydra-battles-single`).

Actually, the issue was more limited than what I thought (not all files are rebuilt). I've reported it as https://github.com/coq-community/coq-nix-toolbox/issues/60.

I still don't understand the issue with `make install-doc` triggering a partial rebuild and making the `hydra-battles-single` job much longer. In the interest of making progress, even if imperfect, I will revert to a version which doesn't build the doc in the `hydra-battles-single` job.

**Comment by Zimmi48** [comment]
This is probably not perfect yet, but seems good enough to merge as is, especially given that I don't have more time to devote to this for now.

**Comment by Casteran** [comment]
Looks to me as a true improvement! Thanks a lot!

Just a (not urgent, I can write an ad-hoc fix) details.
On

**Comment by Casteran** [comment]
Oups ! Here is the full message.




> Le 25 août 2021 à 12:56, Castéran Pierre ***@***.***> a écrit :
> 
> Looks to me as a true improvement! Thanks a lot!
> 
> Just a (not urgent, I can write an ad-hoc fix) details.

On https://coq-community.org/hydra-battles/doc/hydras.pdf <https://coq-community.org/hydra-battles/doc/hydras.pdf> the date on the first page is « January 1, 1980 » since yesterday.



>

**Pull request merged**